src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs

```diff
@@ -0,0 +1,487 @@
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withParsableComplexType_viaMethodConvention_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost("/customers/{customer}/contacts", (Customer customer) => {});
+
+public class Customer
+{
+    public static bool TryParse(string s, IFormatProvider provider, out Customer result)
+    {
+        result = new Customer();
+        return true;
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withHttpContextBindableComplexType_viaImplicitIBindableFromHttp_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+var webApp = WebApplication.Create();
+webApp.MapGet("/customers/{customer}", (Customer customer) => {});
+
+public class Customer : IBindableFromHttpContext<Customer>
+{
+    public static async ValueTask<Customer?> BindAsync(HttpContext context, ParameterInfo parameter)
+    {
+        return new Customer();
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withNullableType_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapGet("/customers/{customer}/contacts", (int? customer) => {});
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withUnbindableComplexType_Fails()
+    {
+        // Arrange
+        var source = $$"""
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ({|#0:[FromBody]Customer customer|}) => {
+    });
+
+public class Customer
+{
+}
+""";
+
+        var expectedDiagnostic = new DiagnosticResult(DiagnosticDescriptors.RouteHandlerParamterComplexTypeIsNotBindable)
+            .WithArguments("customer", "Customer")
+            .WithLocation(0);
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source, expectedDiagnostic);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withInvalidBindAsyncImplementation_Fails()
+    {
+        // Arrange
+        var source = $$"""
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ({|#0:[FromBody]Customer customer|}) => {
+    });
+
+public class Customer
+{
+    public async static Task<Customer> BindAsync(HttpContext context)
+    {
+        return new Customer();
+    }
+}
```

Comment:
This is the case I just mentioned above where this will work, but not the way you want. It will try to deserilize as JSON because of the `[FromBody]` attribute.

We should raise a diagnostic if there is no `[FromBody]` attribute because this returns a `Task` instead of `ValueTask`. It took me a second to figure out why this was supposed to be broken. I wonder if we can do even better than saying `"Parameter '{0}' of type {1} should define a ValueTask<{1}> BindAsync(HttpContext) or implement IBindableFromHttpContext<{1}>."` in this case. I can imagine a lot of people thinking "that's what I did!"

Maybe `"BindAsync must return a ValueTask<{1}> instead of Task<{1}>. Consider implementing IBindableFromHttpContext<{1}>."` for this specific case?

Reply:
Will need another diagnostic but I agree with this. When I was first building out these test cases I got it wrong ;)

Reply:
OK, I've dropped in a more specific diagnostic for this scenario. Took a little while to plumb it together but it means if we want to start getting really helpful with these diagnostics we can.